packages/beacon-node/src/network/reqresp/handlers/beaconBlockAndBlobsSidecarByRoot.ts

```diff
@@ -0,0 +1,77 @@
+import {ContextBytesType, EncodedPayload, EncodedPayloadType} from "@lodestar/reqresp";
+import {eip4844} from "@lodestar/types";
+import {toHex} from "@lodestar/utils";
+import {IBeaconChain} from "../../../chain/index.js";
+import {IBeaconDb} from "../../../db/index.js";
+import {getSlotFromBytes} from "../../../util/multifork.js";
+
+export async function* onBeaconBlockAndBlobsSidecarByRoot(
+  requestBody: eip4844.BeaconBlockAndBlobsSidecarByRootRequest,
+  chain: IBeaconChain,
+  db: IBeaconDb
+): AsyncIterable<EncodedPayload<eip4844.SignedBeaconBlockAndBlobsSidecar>> {
+  const finalizedSlot = chain.forkChoice.getFinalizedBlock().slot;
+
+  for (const blockRoot of requestBody) {
+    const blockRootHex = toHex(blockRoot);
+    const summary = chain.forkChoice.getBlockHex(blockRootHex);
+
+    // NOTE: Only support non-finalized blocks.
+    // SPEC: Clients MUST support requesting blocks and sidecars since the latest finalized epoch.
+    // https://github.com/ethereum/consensus-specs/blob/11a037fd9227e29ee809c9397b09f8cc3383a8c0/specs/eip4844/p2p-interface.md#beaconblockandblobssidecarbyroot-v1
+    if (!summary || summary.slot <= finalizedSlot) {
+      // TODO: Should accept the finalized block? Is the finalized block in the archive DB or hot DB?
+      continue;
+    }
+
+    // finalized block has summary in forkchoice but it stays in blockArchive db
+    const blockBytes = await db.block.getBinary(blockRoot);
```

Comment:
could be `Promise.all()`

Reply:
Will update when the spec for byRoot gets finalized